### PR TITLE
Check permissions in `MessageCorrelationCorrelateProcessor`

### LINKED ISSUE #22885: Verify permissions in `MessageCorrelationCorrelateProcessor`
<h2>Description</h2>

Message correlate is special in the sense that it requires 2 different permission types, depending on if the message creates a PI, or correlates to an existing PI.

Creates a PI:
- **Resource type**: `PROCESS_DEFINITION`
- **Permission type**: `CREATE`
- **Possible permissions**:
    -  `bpmnProcessId - Foo`

Correlated to existing PI:
- **Resource type**: `PROCESS_DEFINITION`
- **Permission type**: `UPDATE`
- **Possible permissions**:
    -  `bpmnProcessId - Foo`

The command is processed for all or for none. If the user is not permitted to correlate the message to one or more of the subscriptions, none of the correlations are correlated and the command should get rejected.

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

It's a bit of an odd one since there's different permissions to check here. The current solution is likely not the most performant, but it is the easiest way to implement it at this time. If it becomes a problem we can improve upon this later.

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #22885

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationCorrelateProcessor.java [hunk 1]
```diff
@@ -138,4 +160,42 @@ private MessageData createMessageData(
         messageCorrelationRecord.getVariablesBuffer(),
         messageCorrelationRecord.getTenantId());
   }
+
+  private Optional<Rejection> isAuthorizedForAllSubscriptions(
```
Comment (nicpuppa) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationCorrelateProcessor.java:164]: 🔧 This method can be refactored a bit:

```
private Optional<Rejection> checkAuthorizationForSubscriptions(
      final TypedRecord<MessageCorrelationRecord> command,
      final Subscriptions correlatingSubscriptions) {

    AuthorizationRequest authRequest = null;

    final var isAuthorized = correlatingSubscriptions.visitSubscriptions(
        subscription -> {

          PermissionType permission = subscription.isStartEventSubscription()
              ? PermissionType.CREATE
              : PermissionType.UPDATE;

          authRequest = new AuthorizationRequest(
              command, AuthorizationResourceType.PROCESS_DEFINITION, permission);
          authRequest.addResourceId(bufferAsString(subscription.getBpmnProcessId()));

          return authCheckBehavior.isAuthorized(authRequest);
        },
        true
    );

    if (!isAuthorized) {
      String errorMessage = UNAUTHORIZED_ERROR_MESSAGE.formatted(
          authRequest.getPermissionType(),
          authRequest.getResourceType()
      );
      return Optional.of(new Rejection(RejectionType.UNAUTHORIZED, errorMessage));
    }

    return Optional.empty();
  }
```

**Review by nicpuppa — APPROVED** [review]
💭 Another approach will be to use the AuthCheck behaviour inside the message correlation behaviour and check the authorization there, in both `correlateToMessageStartEvents` and `correlateToMessageEvents`. In this way we can remove looping again all the subscriptions, and also exit early while looping in the behaviour.
This approach will hide the auth check inside the behaviour of the message correlation, but we can underline this maybe changing the methods name.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationCorrelateProcessor.java (hunk 1 above)
Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationCorrelateProcessor.java:164]: That doesn't work. You cannot assign a value to `authRequest` from within a lambda.

Comment (remcowesterhoud) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationCorrelateProcessor.java:164]: I'll take the permission type part though

Comment (nicpuppa) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageCorrelationCorrelateProcessor.java:164]: Oh yeah, sorry

**Comment by remcowesterhoud** [comment]
> 💭 Another approach will be to use the AuthCheck behaviour inside the message correlation behaviour and check the authorization there, in both correlateToMessageStartEvents and correlateToMessageEvents. In this way we can remove looping again all the subscriptions, and also exit early while looping in the behaviour.
This approach will hide the auth check inside the behaviour of the message correlation, but we can underline this maybe changing the methods name.

I did consider this. The problem is this behavior is used by message publish as well as message correlation. I don't see a very straightforward solution to do this in a nice way currently. Let's revisit if it becomes a problem.

**Pull request merged**